### PR TITLE
[21.05] Pl 130693 ceph luminous

### PR DESCRIPTION
**Pull request by osnyx**
@flyingcircusio/release-managers

## Release process

Impact:

none

Changelog:

* Provide the Ceph Luminuous update using a feature-flag mechanism. (PL-130693)

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)

This is a comprehensive update, however, it doesn't touch too many things from an architectural perspective. We know about some of the deficits of our current implementation (like daemons running as root, but on dedicated machines) that did not change too much.

One aspect of attack surface is covered: the manager HTTP server is bound to localhost instead of the default to listen on all addresses.

Otherwise our key management is a bit simplistic, to allow for smooth automated management and also as things run as root anyway separating keys further seems overkill.

- [x] Security requirements tested? (EVIDENCE)

The whole dev cluster has been updated to this PR's state, individual hosts and Ceph components have been updated to the new version step by step. This is true for both the Ceph storage servers as well as the Ceph client components part of other infrastructure components (kvm, backup).
Major regular maintenance and tasks and operations steps have been replicated on the updated cluster, documentation has been adapted wherever necessary. A stress test (chaosmonkey) has been launched on the dev cluster with no apparent failures.
Several disk format changes, rebuilds and recoverys went through successfully without data loss.
Performance did even increase with the use of Bluestore (beneficial to availability).
Automated NixOS tests

### CODE REVIEW

File: pkgs/overlay.nix [hunk 1]
```diff
@@ -35,10 +35,20 @@ in {
...
   check_megaraid = super.callPackage ./check_megaraid { };
 
-  ceph = (super.callPackage ./ceph {
+  ceph = self.ceph-jewel;
+  ceph-jewel = (super.callPackage ./ceph/jewel {
       pythonPackages = super.python2Packages;
       boost = super.boost155;
   });
+  ceph-luminous = (super.callPackage ./ceph/luminous {
+    # TODO: try newer boost versions
```
Comment (ctheune) [inline pkgs/overlay.nix]: not needed. remove todo marker.

File: pkgs/overlay.nix [hunk 2]
```diff
@@ -414,6 +424,43 @@ in {
...
+      });
+      pecan = thisPy.pkgs.callPackage ./python/pecan { };
+      portend = thisPy.pkgs.callPackage ./python/portend { };
+      pypytools = thisPy.pkgs.callPackage ./python/pypytools { };
+      pyquery = thisPy.pkgs.callPackage ./python/pyquery { };
+      routes = python-super.routes.overrideAttrs (oldattrs: {
+        # work around a weird pythonImportsCheck failure
+        #buildInputs = oldattrs.propagatedBuildInputs;
+        #pythonImportsCheck = [];
+        pythonImportsCheckPhase = ''
```
Comment (ctheune) [inline pkgs/overlay.nix]: I think this can be simplified.

File: pkgs/fc/ceph/default.nix [hunk 3]
```diff
@@ -11,15 +11,15 @@ py.buildPythonApplication rec {
...
-    ceph
     lz4
-    lvm2
     agent
-    utillinux
     util-physical
     python3Packages.requests
   ];
 
+  # FIXME: only temporary
```
Comment (ctheune) [inline pkgs/fc/ceph/default.nix]: well, fix it then ;)

File: pkgs/fc/ceph/src/fc/ceph/main.py [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
 import fc.ceph.logs
 import fc.ceph.maintenance
-import fc.ceph.manage
+import fc.ceph.mgr
+import fc.ceph.mon
+import fc.ceph.osd
+from fc.ceph import Environment
+
+CONFIG_FILE_PATH = Path("/etc/ceph/fc-ceph.conf")
+# FIXME: move to OSD module exported constants
```
Comment (ctheune) [inline pkgs/fc/ceph/src/fc/ceph/main.py]: please check

File: pkgs/fc/ceph/src/fc/ceph/main.py [hunk 5]
```diff
@@ -92,8 +119,17 @@ def main(args=sys.argv[1:]):
...
-        help="Size of journal (LVM size units allowed).",
+        default=fc.ceph.osd.DEFAULT_JOURNAL_SIZE,
+        help="Size of journal (LVM size units allowed). "
+        "Only used if rebuild target is a filestore OSD.",
+    )
+    parser_rebuild.add_argument(
+        "--target-objectstore-type",
+        "-T",
+        choices=OBJECTSTORE_TYPES,
+        # FIXME: do I explicitly need to add the choices to the help text?
```
Comment (ctheune) [inline pkgs/fc/ceph/src/fc/ceph/main.py]: check

File: pkgs/fc/ceph/src/fc/ceph/subsystems.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class VersionedSubsystem(object):
```
Comment (ctheune) [inline pkgs/fc/ceph/src/fc/ceph/subsystems.py]: this seems to  be a duplicate of fc.ceph.VersionedSubsystems. I think this file can just go away.

File: nixos/roles/backyserver.nix [hunk 7]
```diff
@@ -8,6 +8,7 @@ let
   enc = config.flyingcircus.enc;
 
   backy = pkgs.callPackage ../../pkgs/backy.nix { };
+  ceph_common = import ./ceph/ceph-common.nix { inherit lib pkgs; };
```
Comment (ctheune) [inline nixos/roles/backyserver.nix]: please refactor to leverage fclib.ceph instead of introducing a new pattern

File: nixos/roles/ceph/osd.nix [hunk 8]
```diff
@@ -66,6 +71,10 @@ in
           osd op history size = 1000
           osd op history duration = 43200
 
+          # we currently have (too) many PGs in our pools, make sure the clsuter still
```
Comment (ctheune) [inline nixos/roles/ceph/osd.nix]: typo

File: tests/default.nix [hunk 9]
```diff
@@ -48,6 +50,10 @@ in {
   kibana7 = callTest ./kibana.nix { version = "7"; };
   k3s = callTest ./k3s {};
   kvm_host = callTest ./kvm_host.nix {};
```
Comment (ctheune) [inline tests/default.nix]: Suggestin: rename to kvm_host_jewel-ceph-server-jewel and avoid a version-less alias

**Review by ctheune — CHANGES_REQUESTED** [review]

File: nixos/roles/backyserver.nix [hunk 10]
```diff
@@ -22,6 +22,10 @@ let
 
   restoreSingleFiles = pkgs.callPackage ../../pkgs/restore-single-files {};
 
+  backy_rbd_versioned = cephReleaseName: {
```
Comment (dpausp) [inline nixos/roles/backyserver.nix]: style: indentation, camelCase

File: nixos/roles/ceph/osd.nix [hunk 11]
```diff
@@ -120,19 +128,41 @@ in
...
+
+      fc-ceph.settings = let
+        osdSettings =  {
+          release = role.cephRelease;
+          path = fclib.ceph.fc-ceph-path fclib.ceph.releasePkgs.${role.cephRelease};
+        };
+      in {
+        # fc-ceph OSD
+        OSDManager = osdSettings;
+        # FIXME: the module uses the `rbd` binary. While it'd be safer to handle it's
```
Comment (dpausp) [inline nixos/roles/ceph/osd.nix]: remove FIXME

**Review by dpausp — CHANGES_REQUESTED** [review]

**Review by dpausp — APPROVED** [review]

**Pull request merged**